### ISSUE TITLE
Cumulative sum (cumsum) op needed

### ISSUE THREAD

**Issue by langmore** [tensorflow/tensorflow#813]
C.f. [`numpy.cumsum`](http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.cumsum.html)

**Comment by benoitsteiner** [comment]
You can achieve what you're looking for by padding your input with zeros and then convolving the padded tensor with ones. For example, if you have a vector of N elements, you can start by padding it with N-1 elements on the left, and the launch a 1D convolution with a kernel of size N.

Of course that won't be very efficient, but that can get you going for the time being.

**Comment by ibab** [comment]
I'd like to learn how to write ops by implementing this.
I'll work based on the [adding_an_op](https://www.tensorflow.org/versions/0.6.0/how_tos/adding_an_op/index.html) section in the docs.

Can someone give me a few pointers on where this should end up in the tensorflow source tree?
The op itself would probably be registered in [`core/ops/math_ops.cc`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/math_ops.cc).
Should I put the kernel into  a new `core/kernels/cumsum_op.cc` file?

**Comment by girving** [comment]
If we do this, it should probably have a name pattern similar to `reduce_sum`, `reduce_prod`, etc.  Maybe `accumulate_sum`, `accumulate_prod` ([somewhat following numpy](https://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.ufunc.accumulate.html))?  @dave-andersen: Do you have opinions here?

Also, I'm not sure if it should be inclusive or exclusive.

Once we figure that out: yes, the op definitions would go in `core/ops/math_ops.cc`, and the kernel files would have a similar pattern to the reduction ops in `core/kernels`.

**Comment by leconteur** [comment]
I did the following in a personal project. It is probably inefficient, but probably less than a convolution.

`def cumsum(softmax):
    values = tf.split(1, softmax.get_shape()[1], softmax)
    out = []
    prev = tf.zeros_like(values[0])
    for val in values:
        s = prev + val
        out.append(s)
        prev = s
    cumsum = tf.concat(1, out)
    return cumsum`

**Comment by leconteur** [comment]
I did the following in a personal project. It is not very efficient but it's probably a little better than the convolution.

https://gist.github.com/leconteur/82f968f483ddd2234a03

**Comment by langmore** [comment]
Hello ibab.  Have you been able to make any progress on this?  Thanks!

**Comment by ibab** [comment]
Hi, I haven't made a lot of progress yet, but I've spent some time understanding the implementation of the `reduce` kernels.
Writing `accumulate` kernels should be a bit more work, as there's no corresponding function in Eigen that we can delegate the operation to.

**Comment by girving** [comment]
@benoitsteiner: Does eigen have accumulate (scan) capability?

**Comment by ibab** [comment]
I've needed `cumsum` and `cumprod` ops today and wrote the following based on @leconteur's code above:
```python
import tensorflow as tf

def cumsum(xs):
    values = tf.unpack(xs)
    out = []
    prev = tf.zeros_like(values[0])
    for val in values:
        s = prev + val
        out.append(s)
        prev = s
    result = tf.pack(out)
    return result
```
This seems to give me relatively decent speed:
```python
s = tf.Session()
X = np.random.uniform(0, 1, size=(10000, 1000))
v = tf.Variable(X)
s.run(tf.initialize_all_variables())
y = cumsum(v)

 [2]: %timeit s.run(y)
10 loops, best of 3: 110 ms per loop
 [3]: %timeit np.cumsum(X, axis=0)
10 loops, best of 3: 80.3 ms per loop
```

It would still be nice to have dedicated (and parallel?) ops for this, though.
Maybe @benoitsteiner can comment on whether adding an `accumulate` implementation to Eigen would make sense. I might have time to contribute that.

**Comment by langmore** [comment]
I don't know the tensorflow internals, but it seems to me that this would be very slow.  Your for loop adds 10,000 nodes to the graph, and the cumsum evaluation requires evaluation of 10,000 separate nodes, rather than one node that calls into an optimized loop.  Each node adds two length 1000 vectors, and this is fast, but I don't imagine iterating through 10,000 nodes is fast.  I must be misunderstanding, how?

**Comment by ibab** [comment]
I've changed the shape to `(10000,)` to isolate the overhead from calling that many nodes a bit better.
Results:
``` 
[2]: %timeit s.run(y)
10 loops, best of 3: 36.6 ms per loop
 [3]: %timeit np.cumsum(X, axis=0)
The slowest run took 8.94 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 28.2 µs per loop
```
So it seems that although there's considerable overhead, it's smaller than the time required to add the `1000` vectors above.

**Comment by benoitsteiner** [comment]
@ibab @girving Eigen currently doesn't provide the functionality needed to efficiently implement an  accumulate op. You're more than welcome to contribute the code to Eigen, that would be a very useful extension.

**Comment by ibab** [comment]
Thanks, @benoitsteiner!
I assume that I could structure it in a similar way to [`TensorReduction.h`](https://bitbucket.org/eigen/eigen/src/59859fde3bc7173175087ed5d002389b4e76876f/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h?at=default&fileviewer=file-view-default), except that `accumulate` would preserve the original shape of the tensor.
This might turn out to be too difficult for me, but I'll give it a try over the next few weeks.

**Comment by chemelnucfin** [comment]
Does this article help?
http://http.developer.nvidia.com/GPUGems3/gpugems3_ch39.html

**Comment by benoitsteiner** [comment]
@ibab The TensorReduction.h code is very optimized and therefore pretty complex. A simpler operation that you could use as a template is TensorReverse.h (which reverse the order of the coefficients along one or more dimensions) or TensorShuffling.h (which reorders the dimension - this is a transposition in the 2d case)

**Comment by ibab** [comment]
I've submitted a [PR to the Eigen repo](https://bitbucket.org/eigen/eigen/pull-requests/189/add-scan-op-to-tensor-module) containing a simple single-threaded implementation of a `TensorScanOp`, which has just been merged.
Ideally I'd like to extend this with thread-based and a CUDA implementations using an approach like in the article that @chemelnucfin quoted above.
Let's see if I'll have enough time to do it.
Failing that, I think that adding TensorFlow ops that use the naive implementation would still be useful.

**Comment by girving** [comment]
@benoitsteiner: How does the merging / syncing process work with external Eigen changes?  Do we have to wait for any period of time before using them in TensorFlow?

**Comment by benoitsteiner** [comment]
@girving We can update the bazel build and workspace files to pull the latest version of Eigen any time. I typically wait for about one day before doing so though, since that gives me a change to check that the Eigen nightly regressions tests are clean before updating TensorFlow.

**Comment by girving** [comment]
The product version of this would be useful to fix https://github.com/tensorflow/tensorflow/issues/2641.

**Comment by ibab** [comment]
This has been fixed by #2711, so this issue can be closed.

**Comment by vrv** [comment]
Yup, thanks!!

**Comment by kbrems** [comment]
As a potential alternative, we have implemented a python-only framework for adding custom operators to tensorflow, the Operator Vectorization Library: 
https://github.com/opveclib/opveclib. Multi-threaded C++ and CUDA code is automatically generated from the python operators, which can be used in a Tensorflow application with only a binary install of tensorflow. We have implemented cumsum and cumprod operators as an example and benchmarks against the new Tensorflow/Eigen scan operators on the original example in this issue are favorable - test code at: https://github.com/opveclib/opveclib/blob/master/opveclib/examples/test_accumulate.py

input: X = np.random.uniform(0, 1, size=(10000,1000))
Machine: 48 CPU cores
GPU - GeForce GTX TITAN
DEBUG:opveclib:Best numpy time (ms): 148.728299141
DEBUG:opveclib:Best ovl cpp time (ms): 41.827047
DEBUG:opveclib:Best ovl cuda time  (ms): 4.738103
DEBUG:opveclib:Best tf + ovl time  (ms) on /cpu:0 :62.8859996796
DEBUG:opveclib:Best tf cumsum time  (ms) on /cpu:0 :141.862869263
DEBUG:opveclib:Best tf + ovl time  (ms) on /gpu:0 :4.96506690979
DEBUG:opveclib:Best tf cumsum time  (ms) on /gpu:0 :9.82308387756

**Comment by ibab** [comment]
opveclib is very cool!
Note that both the CPU and GPU implementation that I've added to Eigen are extremely naive.
I've been planning to add optimized implementations, but haven't managed to get around to it yet.
It would be interesting to test the performance again with the optimized implementations.
